### PR TITLE
Add missing validDuration and maxFee columns

### LINKED ISSUE #221: DB Transactions: t_transactions table missing validDuration+MaxFee
**Detailed Description**
The `t_transactions` table doesn't contain a column to store the valid duration and maxFee of a transaction.

- Database update to the `t_transactions` table to include `valid_duration` and `max_fee` fields to be derived from the [protobuf](https://github.com/hashgraph/hedera-protobuf/blob/master/src/main/proto/TransactionBody.proto) TransactionBody.transactionFee and transactionValidDuration
- java updates to RecordFileLogger.storeRecord() to store these fields (`sqlInsertTransaction` changes).
- Update integration and unit tests

**Actual Behavior**
Inspect `t_transaction` table, witness lack of columns.

**Expected Behavior**
A transaction's valid duration and maxFee should be recorded in the table.

### LINKED ISSUE #337: Inconsistent nanoseconds conversion in REST API transaction.js transfer list creation causing 404 on single transaction GET
**Detailed Description**
transactions.js createTransferLists() converts valid_start_timestamp nanoseconds into a seconds.nnnnnnnnn format since epoch. This logic is done twice to populate valid_start_timestamp and as part of the transaction_id creation.
The logic used between the two was inconsistent and was resulting in the transaction_id having a seconds.nnnnnnnnn value that differed from the valid_start_timestamp.

The result was api calls to /transactions/<transaction_id> were returning a 404 as the entity could not be found.
This seems to be a regression that has yet to hit test or main net.

**Actual Behavior**
1. Make a /api/v1/transactions call.
2. Take a transaction_id from any returned transaction
3. Make a /api/v1/transactions/<transaction_id> call to get the single transaction
4. A 404 'Not found' is returned by rest-api

**Expected Behavior**
A valid json object of the single transaction matching the transaction_id from the /transactions call should be returned

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
This changes adds the missing maxFee and validDuration columns from a transaction to the transactions table in the database.
It also exposes these fields as parameters in the rest-api transaction object results.
It also fixes an inconsistency in how nanoseconds were being converted for the valid_start_time.

Signed-off-by: Nana-EC <56320167+Nana-EC@users.noreply.github.com>

Changes include
- New flywaydb migration version for sql to alter the Postgres transactions table
- Updated to the Transaction domain and Transactions objects in tests with the matching change
- Update to RecordFileLogger sqlinserttransaction calls to set new columns

**Which issue(s) this PR fixes**:
Fixes #221 
Fixes #337 

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: src/main/java/com/hedera/mirror/domain/Transaction.java [hunk 1]
```diff
@@ -1,5 +1,10 @@
 package com.hedera.mirror.domain;
 
+import javax.persistence.Column;
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/mirror/domain/Transaction.java]: import statements should be below license header

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java [hunk 2]
```diff
@@ -288,6 +307,7 @@ public static void storeRecord(Transaction transaction, TransactionRecord txReco
 		sqlInsertTransaction.setLong(F_TRANSACTION.VALID_START_NS.ordinal(), validStartNs);
 		sqlInsertTransaction.setInt(F_TRANSACTION.FK_TRANS_TYPE_ID.ordinal(), getTransactionTypeId(body));
 		sqlInsertTransaction.setLong(F_TRANSACTION.FK_REC_FILE_ID.ordinal(), fileId);
+		sqlInsertTransaction.setLong(F_TRANSACTION.VALID_DURATION.ordinal(), validDuration.getSeconds());
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Do we need to check `body.hasTransactionValidDuration()` first?

File: src/test/java/com/hedera/mirror/repository/TransactionRepositoryTest.java [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package com.hedera.mirror.repository;
 
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (steven-sheehy) [inline src/test/java/com/hedera/mirror/repository/TransactionRepositoryTest.java]: not sure why this was moved. should be below license header

File: src/main/java/com/hedera/mirror/domain/Transaction.java (hunk 1 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/mirror/domain/Transaction.java]: IDE did this, will move it back down

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 2 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: no, the logic in body.getTransactionValidDuration() does the same check that body.hasTransactionValidDuration() does and return a default duration in the negative case

File: src/test/java/com/hedera/mirror/repository/TransactionRepositoryTest.java (hunk 3 above)
Comment (Nana-EC) [reply src/test/java/com/hedera/mirror/repository/TransactionRepositoryTest.java]: Same issue as above IDE moved it up, fixing it.

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
+--
```
Comment (mike-burrage-hedera) [inline src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: Make filename version V1.12.0 (we've been updating this minor version every minor release).

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+ -- Add validDuration and maxFee columns to t_transactions
+ --
+
+ alter table t_transactions
+     add column valid_duration bigint null;
+
+ alter table t_transactions
+     add column max_fee bigint null;
```
Comment (mike-burrage-hedera) [inline src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: datatype `hbar_tinybars` instead of `bigint`. For clarity

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql [hunk 6]
```diff
@@ -0,0 +1,9 @@
+--
+ -- Add validDuration and maxFee columns to t_transactions
+ --
+
+ alter table t_transactions
+     add column valid_duration bigint null;
```
Comment (mike-burrage-hedera) [inline src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: Rename to `valid_duration_seconds`. For clarity

File: src/main/java/com/hedera/mirror/domain/Transaction.java [hunk 7]
```diff
@@ -58,4 +61,7 @@
 
     private Long validStartNs;
 
+    private Long validDuration;
```
Comment (mike-burrage-hedera) [inline src/main/java/com/hedera/mirror/domain/Transaction.java]: validDurationSeconds

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 2 above)
Comment (gregscullard) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: If duration was a scalar (like validDuration.getSeconds()), I'd argue we don't need to since the value would default to 0 per protobuf.
Not 100% sure what happens if there is no duration in this instance since it's a message. It may be worth checking and setting a duration long variable to 0 if missing.

long transactionValidDuration = 0;
if (body.hasTransactionValidDuration) {
  transactionValidDuration = body.getTransactionValidDuration.getSeconds();
}
...
...
sqlInsertTransaction.setLong(F_TRANSACTION.VALID_DURATION.ordinal(), transactionValidDuration);

For example.

File: rest-api/transactions.js [hunk 8]
```diff
@@ -48,6 +48,8 @@ const createTransferLists = function (rows, arr) {
             transactions[row.consensus_ns]['result'] = row['result'];
             transactions[row.consensus_ns]['result'] = row['result'];
             transactions[row.consensus_ns]['name'] = row['name'];
+            transactions[row.consensus_ns]['max_fee'] = Number(row['max_fee']);
+            transactions[row.consensus_ns]['valid_duration'] = Number(row['valid_duration']);
```
Comment (mike-burrage-hedera) [inline rest-api/transactions.js]: Both of these fields should be set to null if req['xyz'] is null. Implemented as above will return "0" to the client in cases where we didn't actually record the max_fee or valid_duration (records processed and existing in the database before is db/java change was implemented).

File: rest-api/__tests__/integration.test.js [hunk 9]
```diff
@@ -169,8 +169,8 @@ const addAccount = async function(accountId) {
 
 const addTransaction = async function(consensusTimestamp, fileId, payerAccountId, transfers) {
     await sqlConnection.query(
-        'insert into t_transactions (consensus_ns, valid_start_ns, fk_rec_file_id, fk_payer_acc_id, fk_node_acc_id, fk_result_id, fk_trans_type_id) values ($1, $2, $3, $4, $5, $6, $7);',
-        [consensusTimestamp, consensusTimestamp - 1, fileId, accountEntityIds[payerAccountId], accountEntityIds[2], 24, 2]);
+        'insert into t_transactions (consensus_ns, valid_start_ns, fk_rec_file_id, fk_payer_acc_id, fk_node_acc_id, fk_result_id, fk_trans_type_id, valid_duration, max_fee) values ($1, $2, $3, $4, $5, $6, $7, $8, $9);',
+         [consensusTimestamp, consensusTimestamp - 1, fileId, accountEntityIds[payerAccountId], accountEntityIds[2], 24, 2, 11, 33]);
```
Comment (mike-burrage-hedera) [inline rest-api/__tests__/integration.test.js]: Add a test case for both fields being null.

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql (hunk 6 above)
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: Add "if not exists" to column additions. We should be standardizing on always doing this.
`add column if not exists ...`

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 2 above)
Comment (mike-burrage-hedera) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: I'd set it to null (in the DB) `if !hasTransactionValidDuration`. It _shouldn't_ get through preconsensus checks with this field missing, but it's possible.

File: rest-api/__tests__/integration.test.js (hunk 9 above)
Comment (Nana-EC) [reply rest-api/__tests__/integration.test.js]: Good idea. Done

File: rest-api/transactions.js (hunk 8 above)
Comment (Nana-EC) [reply rest-api/transactions.js]: Added a util function to take a value and return the Number value or null if it's not truthy

File: src/main/java/com/hedera/mirror/domain/Transaction.java (hunk 7 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/mirror/domain/Transaction.java]: Done

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 2 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Used hasTransactionValidDuration and where false set it to null or took the seconds if false

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql (hunk 6 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: Added

File: src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql (hunk 5 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.11.7__transactions_maxfee_duration.sql]: Changed

File: rest-api/utils.js [hunk 10]
```diff
@@ -593,6 +593,15 @@ const encodeBase64 = function (buffer) {
...
 }
 
+/**
+ * 
+ * @param {Number} num Nullable number
+ * @returns {Any} value of paramter or null if null 
+ */
+const getNullableNumber = function(num) {
+    return num ? Number(num) : null;
+}
```
Comment (mike-burrage-hedera) [inline rest-api/utils.js]: This returns null for 0.
The goal should be to return null for null, and otherwise a number.
Also, with a new function added - there need to be unit tests for this (including null, 0, and some positive number).

Comment (Nana-EC) [reply rest-api/utils.js]: Good catch. Forgot 0 was Falsy for numbers

Comment (Nana-EC) [reply rest-api/utils.js]: Fixed and it's added

**Comment by Nana-EC** [comment]
> Can you
> 
> * Add these two fields to AbstractRecordFileLoggerTest.assertTransaction, this test checks that a transaction is logged correctly in the database
> * Update the CHANGELOG to document API and DB changes
> * Check with Simi if documentation on docs.hedera.com needs to be updated with these changes

Did 1 & 2, waiting on reply for Simi on this.

File: rest-api/__tests__/testutils.test.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+'use strict';
+
+const request = require('supertest');
+const utils = require('../utils.js');
```
Comment (steven-sheehy) [inline rest-api/__tests__/testutils.test.js]: I was confused and thought this class contained test utils, not that it was testing utils.js. Even more confusing since there is already testutils.js. Might be better to name utils.test.js.

File: rest-api/transactions.js [hunk 12]
```diff
@@ -48,16 +48,13 @@ const createTransferLists = function (rows, arr) {
...
                 row.node_shard + '.' + row.node_realm + '.' + row.node_num;
 
             // Construct a transaction id using format: shard.realm.num-sssssssssss-nnnnnnnnn
-            transactions[row.consensus_ns]['transaction_id'] =
-                row.entity_shard + '.' +
-                row.entity_realm + '.' +
-                row.entity_num + '-' +
-                Math.floor(validStartTimestamp / billion) + '-' +
-                String(validStartTimestamp % billion).padStart(9, '0');
+            transactions[row.consensus_ns]['transaction_id'] = utils.createTransactionId(row.entity_shard,row.entity_realm,row.entity_num,validStartTimestamp);
```
Comment (steven-sheehy) [inline rest-api/transactions.js]: Nice bugfix, but you should document reasoning in PR description. Also, billion is now unused and should be deleted.

File: rest-api/utils.js [hunk 13]
```diff
@@ -593,6 +602,27 @@ const encodeBase64 = function (buffer) {
...
 }
 
+/**
+ * 
+ * @param {Any} num Nullable number
+ * @returns {Any} Number value of parameter or null if null 
+ */
+const getNullableNumber = function(num) {
+    return num == null ? null : Number(num);
+}
```
Comment (mike-burrage-hedera) [inline rest-api/utils.js]: Since max_fee and valid_duration_seconds are bigints in the DB, shouldn't this be a math.bignumber to avoid losing precision?

File: rest-api/__tests__/testutils.test.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    test('Verify getNullableNumber returns correct result for undefined', () => {
+        var val = utils.getNullableNumber(undefined);
+        expect(val).toBe(null);
+    });
+
+    test('Verify getNullableNumber returns correct result for valid number', () => {
+        var validNumber = 10;
+        var val = utils.getNullableNumber(validNumber);
+        expect(val).toBe(validNumber);
+    });
```
Comment (mike-burrage-hedera) [inline rest-api/__tests__/testutils.test.js]: What about a full use of the bigint's precision? Something like '9223372036854775807'. Shouldn't lose precision via this function call.

File: rest-api/__tests__/testutils.test.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+    test('Verify nsToSecNsWithHyphen returns correct result for null validStartNs', () => {
+        var val = utils.nsToSecNsWithHyphen(null);
+        expect(val).toBe('0-000000000');
+    });
+});
+
+describe('Utils createTransactionId tests', () => {
+    var validStartNs = 1568411656265684000;
```
Comment (mike-burrage-hedera) [inline rest-api/__tests__/testutils.test.js]: I think making validStartNs either a string or math.bignumber and replacing the last 3 digits with non-zeros, to test out precision handling would be better. The number above is at the limit of precision for a Number as-is.

File: rest-api/__tests__/testutils.test.js (hunk 11 above)
Comment (Nana-EC) [reply rest-api/__tests__/testutils.test.js]: Can do. The other class doesn't seem to actually test utils. I see it providing functions that other test classes utilize.

File: rest-api/transactions.js (hunk 12 above)
Comment (Nana-EC) [reply rest-api/transactions.js]: Yeah gonna create an issue to capture this. Will remove billion also

File: rest-api/transactions.js [hunk 16]
```diff
@@ -327,14 +326,16 @@ const getOneTransaction = function (req, res) {
...
-        if (ret.transactions.length === 0) {
-            res.status(404)
-                .send('Not found');
-            return;
-        }
+        // if (ret.transactions.length === 0) {
+        //     res.status(404)
+        //         .send('Not found');
+        //     return;
+        // }
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply rest-api/transactions.js]: Mistake, didn't mean to push this. Reverting

File: CHANGELOG.md [hunk 17]
```diff
@@ -2,6 +2,12 @@
 
 Besides bug fixes, some features may have changed with this release which need your attention, these will be listed here.
 
+### Addded `valid_duration_seconds` and `max_fee` columns on `t_transactions`
+
+`valid_duration_seconds` represents the seconds for which a submitted transaction is to be deemed valid beyond the start time. The transaction is invalid if consensusTimestamp > transactionValidStart + `valid_duration_seconds`
+`max_fee`represents the maximum transaction fee the client is willing to pay, which is split between the network and the node
+Both values are set to null if a valid value is not received.
+
```
Comment (gregscullard) [inline CHANGELOG.md]: Suggest you also refer changes to REST api responses.

File: rest-api/__tests__/testutils.test.js (hunk 14 above)
Comment (Nana-EC) [reply rest-api/__tests__/testutils.test.js]: Done

File: rest-api/__tests__/testutils.test.js (hunk 15 above)
Comment (Nana-EC) [reply rest-api/__tests__/testutils.test.js]: Done
